lib/mayaUsd/render/vp2RenderDelegate/proxyRenderDelegate.cpp

```diff
@@ -699,6 +702,159 @@ void ProxyRenderDelegate::_UpdateSelectionStates()
+    bool renderPurposeChanged, proxyPurposeChanged, guidePurposeChanged;
+    _proxyShapeData->UpdatePurpose(
+        &renderPurposeChanged, &proxyPurposeChanged, &guidePurposeChanged);
+    if (renderPurposeChanged || proxyPurposeChanged || guidePurposeChanged)
+    {
+        // Build the list of render tags which were added or removed (changed)
+        // and the list of render tags which were removed.
+        TfTokenVector changedRenderTags;
+        TfTokenVector removedRenderTags;
+        if (renderPurposeChanged) {
+            changedRenderTags.push_back(HdRenderTagTokens->render);
+            if (!_proxyShapeData->DrawRenderPurpose())
+                removedRenderTags.push_back(HdRenderTagTokens->render);
+        }
+        if (proxyPurposeChanged) {
+            changedRenderTags.push_back(HdRenderTagTokens->proxy);
+            if (!_proxyShapeData->DrawProxyPurpose())
+                removedRenderTags.push_back(HdRenderTagTokens->proxy);
+        }
+        if (guidePurposeChanged) {
+            changedRenderTags.push_back(HdRenderTagTokens->guide);
+            if (!_proxyShapeData->DrawGuidePurpose())
+                removedRenderTags.push_back(HdRenderTagTokens->guide);
+        }
+
+        // Mark all the rprims which have a render tag which changed dirty
+        SdfPathVector    rprimsToDirty = _GetFilteredRprims(*(_defaultCollection.get()), changedRenderTags);
+        HdChangeTracker& changeTracker = _renderIndex->GetChangeTracker();
+        for (auto id : rprimsToDirty) {
+            changeTracker.MarkRprimDirty(id, HdChangeTracker::DirtyRenderTag);
+        }
+
+        // Make a special pass over the removed render tags so that the objects
+        // can hide themselves.
+        _taskController->SetRenderTags(removedRenderTags);
+        _engine.Execute(_renderIndex.get(), &_dummyTasks);
+
+        // Set the new render tags correctly.The regular execute will cause
+        // any newly visible rprims will update and mark themselves visible.
+        TfTokenVector renderTags
+            = { HdRenderTagTokens->geometry }; // always draw geometry render tag purpose.
+        if (_proxyShapeData->DrawRenderPurpose()) {
+            renderTags.push_back(HdRenderTagTokens->render);
+        }
+        if (_proxyShapeData->DrawProxyPurpose()) {
+            renderTags.push_back(HdRenderTagTokens->proxy);
+        }
+        if (_proxyShapeData->DrawGuidePurpose()) {
+            renderTags.push_back(HdRenderTagTokens->guide);
+        }
+        _taskController->SetRenderTags(renderTags);
+    }
+}
+
+
+// Copied from renderIndex.cpp, the code that does HdRenderIndex::GetDrawItems. But I just want the rprimIds, I don't want to go all the way to draw items.
+struct _FilterParam {
+    const HdRprimCollection& collection;
+    const TfTokenVector&     renderTags;
+    const HdRenderIndex*     renderIndex;
+};
+
+static bool _DrawItemFilterPredicate(const SdfPath& rprimID, const void* predicateParam)
```

Comment:
How about we put this method in anonymous namespace we already have at the top of this file? In such case, the static keyword can be ignored.